core/src/core/pmmr/segment.rs

```diff
@@ -0,0 +1,519 @@
+// Copyright 2020 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Segment of a PMMR.
+
+use crate::core::hash::Hash;
+use crate::core::pmmr::{self, Backend, ReadablePMMR, ReadonlyPMMR};
+use crate::ser::{Error, PMMRIndexHashable, PMMRable, Readable, Reader, Writeable, Writer};
+use croaring::Bitmap;
+use std::cmp::min;
+use std::fmt::{self, Debug};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+/// Error related to segment creation or validation
+pub enum SegmentError {
+	/// An expected leaf was missing
+	MissingLeaf(u64),
+	/// An expected hash was missing
+	MissingHash(u64),
+	/// The segment does not exist
+	NonExistent,
+	/// The segment is fully pruned
+	Pruned,
+	/// Mismatch between expected and actual root hash
+	Mismatch,
+}
+
+impl fmt::Display for SegmentError {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		match self {
+			SegmentError::MissingLeaf(idx) => write!(f, "Missing leaf at pos {}", idx),
+			SegmentError::MissingHash(idx) => write!(f, "Missing hash at pos {}", idx),
+			SegmentError::NonExistent => write!(f, "Segment does not exist"),
+			SegmentError::Pruned => write!(f, "Segment is fully pruned"),
+			SegmentError::Mismatch => write!(f, "Root hash mismatch"),
+		}
+	}
+}
+
+/// Tuple that defines a segment of a given PMMR
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct SegmentIdentifier {
+	/// 2-Log of the size of a segment
+	pub log_size: u8,
+	/// Zero-based index of the segment
+	pub idx: u64,
+}
+
+impl Readable for SegmentIdentifier {
+	fn read<R: Reader>(reader: &mut R) -> Result<Self, Error> {
+		let log_size = reader.read_u8()?;
+		let idx = reader.read_u64()?;
+		Ok(Self { log_size, idx })
+	}
+}
+
+impl Writeable for SegmentIdentifier {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), Error> {
+		writer.write_u8(self.log_size)?;
+		writer.write_u64(self.idx)
+	}
+}
+
+/// Segment of a PMMR: unpruned leaves and the necessary data to verify
+/// segment membership in the original MMR.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Segment<T> {
+	identifier: SegmentIdentifier,
+	hashes: Vec<(u64, Hash)>,
+	leaf_data: Vec<(u64, T)>,
+	proof: SegmentProof,
+}
+
+impl<T> Segment<T> {
+	/// Maximum number of leaves in a segment, given by `2**b`
+	#[inline]
+	fn segment_capacity(&self) -> u64 {
+		1 << self.identifier.log_size
+	}
+
+	/// Offset (in leaf idx) of first leaf in the segment
+	#[inline]
```

Comment:
Do we need `#[inline]` everywhere? I'm not clear on guidelines for explicitly doing this.

Reply:
My assumption was always that it should be added for "sufficiently small" functions to hint to the compiler it should be inlined, but I did some googling and apparently it only has an effect across crate boundaries. Since these functions are internal to the segment i will remove the inlining here.